File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
import sys

input = sys.stdin.readline
sys.setrecursionlimit(10 ** 5)

def dfs(start):
    visit[start] = True

    for i in maps[start]:
        if not visit[i]:
            ans[i] = start
            dfs(i)

n = int(input())

maps = [[] for _ in range(n + 1)]
visit = [False] * (n + 1)
ans = [0] * (n + 1)
for _ in range(n - 1):
    a, b = map(int, input().split())
    maps[a].append(b)
    maps[b].append(a)

dfs(1)
for i in range(2, n + 1):
    print(ans[i])

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.java
import java.io.*;
import java.util.*;

class Main {
    static int n;
    public static int [] ans;
    public static boolean [] visit;
    public static ArrayList<Integer>[] maps;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());
        maps = new ArrayList[n + 1];
        visit = new boolean[n + 2];
        ans = new int [n + 2];

        for (int i = 1; i < n + 1; i++) {
            maps[i] = new ArrayList<Integer>();
        }

        for (int i = 0; i < n - 1; i++) {
            String [] nums = br.readLine().split(" ");
            int a = Integer.parseInt(nums[0]);
            int b = Integer.parseInt(nums[1]);

            maps[a].add(b);
            maps[b].add(a);
        }
        DFS(1);

        for (int i = 2; i < n + 1; i++) System.out.println(ans[i]);
    }
    private static void DFS(int start) {
        visit[start] = true;
        
        for (int i: maps[start]) {
            if (!visit[i]) {
                ans[i] = start;
                DFS(i);
            }
        }
    }
}